File: app/src/main/java/tpcreative/co/qrscanner/ui/filecolor/ChangeFileColorActivity.java
package tpcreative.co.qrscanner.ui.filecolor;
import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.widget.ImageView;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.gson.Gson;
import com.google.zxing.BarcodeFormat;
import com.google.zxing.EncodeHintType;
import com.journeyapps.barcodescanner.BarcodeEncoder;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import butterknife.BindView;
import tpcreative.co.qrscanner.R;
import tpcreative.co.qrscanner.common.SettingsSingleton;
import tpcreative.co.qrscanner.common.activity.BaseActivitySlide;
import tpcreative.co.qrscanner.common.controller.PrefsController;
import tpcreative.co.qrscanner.common.presenter.BaseView;
import tpcreative.co.qrscanner.common.view.GridSpacingItemDecoration;
import tpcreative.co.qrscanner.model.EnumStatus;
import tpcreative.co.qrscanner.model.Theme;

public class ChangeFileColorActivity extends BaseActivitySlide implements BaseView  ,ChangeFileColorAdapter.ItemSelectedListener{

    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;
    @BindView(R.id.imgResult)
    ImageView imgResult;
    private Bitmap bitmap;
    private ChangeFileColorPresenter presenter;
    private ChangeFileColorAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chage_file_color);
        final Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        initRecycleView(getLayoutInflater());
        presenter = new ChangeFileColorPresenter();
        presenter.bindView(this);
        presenter.getData();
    }

    public void initRecycleView(LayoutInflater layoutInflater) {
        adapter = new ChangeFileColorAdapter(layoutInflater, getApplicationContext(), this);
        RecyclerView.LayoutManager mLayoutManager = new GridLayoutManager(getApplicationContext(), 4);
        recyclerView.setLayoutManager(mLayoutManager);
        recyclerView.addItemDecoration(new GridSpacingItemDecoration(4, 4, true));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setAdapter(adapter);
    }

    @Override
    public void onClickItem(int position) {
        presenter.mTheme = presenter.mList.get(position);
        PrefsController.putString(getString(R.string.key_theme_object),new Gson().toJson(presenter.mTheme));
        presenter.getData();
        SettingsSingleton.getInstance().onUpdated();
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home: {
                onBackPressed();
                return true;
            }
        }
        return false;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void onStartLoading(EnumStatus status) {

    }

    @Override
    public void onStopLoading(EnumStatus status) {

    }

    @Override
    public void onError(String message, EnumStatus status) {

    }

    @Override
    public void onError(String message) {

    }

    @Override
    public void onSuccessful(String message) {

    }

    @Override
    public void onSuccessful(String message, EnumStatus status) {
        switch (status){
            case SHOW_DATA:{
                adapter.setDataSource(presenter.mList);
                onGenerateReview("123");
                break;
            }
        }
    }

    @Override
    public Activity getActivity() {
        return this;
    }

    @Override
    public Context getContext() {
        return getApplicationContext();
    }


    @Override
    public void onSuccessful(String message, EnumStatus status, Object object) {

    }

    @Override
    public void onSuccessful(String message, EnumStatus status, List list) {

    }

    public void onGenerateReview(String code){
        try {
            BarcodeEncoder barcodeEncoder = new BarcodeEncoder();
            Map<EncodeHintType, Object> hints = new EnumMap<>(EncodeHintType.class);
            hints.put(EncodeHintType.MARGIN, 2);
            Theme theme = Theme.getInstance().getThemeInfo();
            bitmap = barcodeEncoder.encodeBitmap(this,theme.getPrimaryDarkColor(),code, BarcodeFormat.QR_CODE, 100, 100,hints);
            imgResult.setImageBitmap(bitmap);
        } catch(Exception e) {
            e.printStackTrace();
        }
    }

}

File: app/src/main/java/tpcreative/co/qrscanner/ui/filecolor/ChangeFileColorActivity.kt
package tpcreative.co.qrscanner.ui.filecolor
import android.graphics.Bitmap
import android.os.Bundle
import android.view.*
import android.widget.LinearLayout
import tpcreative.co.qrscanner.ui.scanner.cpp.BarcodeEncoder
import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
//import com.journeyapps.barcodescanner.BarcodeEncoder
import tpcreative.co.qrscanner.common.Constant
import tpcreative.co.qrscanner.common.SettingsSingleton
import tpcreative.co.qrscanner.common.Utils
import tpcreative.co.qrscanner.common.activity.BaseActivitySlide
import tpcreative.co.qrscanner.common.services.QRScannerApplication
import tpcreative.co.qrscanner.common.view.ads.AdsView
import tpcreative.co.qrscanner.databinding.ActivityChageFileColorBinding
import tpcreative.co.qrscanner.model.*
import java.util.*

class ChangeFileColorActivity : BaseActivitySlide(), ChangeFileColorAdapter.ItemSelectedListener {
    private var bitmap: Bitmap? = null
    lateinit var viewModel : ChangeFileColorViewModel
    var adapter: ChangeFileColorAdapter? = null
    lateinit var llSmallAds : AdsView
    lateinit var binding : ActivityChageFileColorBinding
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityChageFileColorBinding.inflate(layoutInflater)
        setContentView(binding.root)
        llSmallAds = AdsView(this)
        initUI()

    }

    override fun onClickItem(position: Int) {
        viewModel.mTheme = dataSource[position]
        Utils.setQRCodeThemePosition(position)
        getData()
        SettingsSingleton.getInstance()?.onUpdated()
    }

    override fun onResume() {
        super.onResume()
        QRScannerApplication.getInstance().onResumeAds(EnumScreens.CHANGE_COLOR_SMALL)
        QRScannerApplication.getInstance().onResumeAds(EnumScreens.CHANGE_COLOR_LARGE)
    }

    override fun onPause() {
        QRScannerApplication.getInstance().onPauseAds(EnumScreens.CHANGE_COLOR_SMALL)
        QRScannerApplication.getInstance().onPauseAds(EnumScreens.CHANGE_COLOR_LARGE)
        super.onPause()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                showAds()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    /*show ads*/
    fun doShowAds(isShow: Boolean) {
        if (isShow) {
            QRScannerApplication.getInstance().loadChangeColorSmallView(llSmallAds.getSmallAds())
            QRScannerApplication.getInstance().loadChangeColorLargeView(binding.llLargeAds)
        }
    }

    fun onGenerateReview(code: String?) {
        try {
            val barcodeEncoder = BarcodeEncoder()
            val hints: MutableMap<EncodeHintType?, Any?> = EnumMap(EncodeHintType::class.java)
            hints[EncodeHintType.MARGIN] = 2
            val theme: Theme? = Theme.getInstance()?.getThemeInfo()
            bitmap = barcodeEncoder.encodeBitmap(this, theme?.getPrimaryDarkColor() ?:0, code, BarcodeFormat.QR_CODE, Constant.QRCodeViewWidth, Constant.QRCodeViewHeight, hints)
            binding.imgResult.setImageBitmap(bitmap)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private val dataSource : MutableList<Theme>
        get() {
            return adapter?.getDataSource() ?: mutableListOf()
        }
}